plugin/storage/memory/memory.go

```diff
@@ -339,3 +331,68 @@ func flattenTags(span *model.Span) model.KeyValues {
 	}
 	return retMe
 }
+
+func isLeaf(span *model.Span, spans []*model.Span) bool {
+	for _, s := range spans {
+		for _, ref := range s.References {
+			if ref.RefType == model.ChildOf && ref.SpanID == span.SpanID {
+				return false // Not a leaf if it's the parent of another span
+			}
+		}
+	}
+	return true
+}
+
+func updateServiceDependencyLinks(parentService, childService string, deps map[string]*model.DependencyLink) {
+	depKey := parentService + "&&&" + childService
+	if _, ok := deps[depKey]; !ok {
+		deps[depKey] = &model.DependencyLink{
+			Parent:    parentService,
+			Child:     childService,
+			CallCount: 1,
+		}
+	} else {
+		deps[depKey].CallCount++
+	}
+}
+
+func inferServiceName(span *model.Span, isLeaf bool) string {
+	serviceName := "inferred"
```

Comment:
this doesn't make a lot of sense in many cases. E.g. if you have `peer.service`, then why should you use `inferred-` prefix? If that `peer.service` value is actually consistent across the architecture (i.e. there may be proper spans from this service in other traces), then using the same name will reuse the same node, which is better representation of the service map than splitting into `interred-` + another

Reply:
I see your point about the "inferred-" prefix not always making sense, especially when we already have a clear identifier like `peer.service`. It's true that using the `peer.service` tag directly would be more consistent and give us a better picture of the service interactions. I'll adjust the code to use `peer.service` as the service name without the "inferred-" prefix when it's available. This should help us avoid creating unnecessary distinctions in the service map.

I propose the following adjustment to the `inferServiceName` function:

```
func inferServiceName(span *model.Span, isLeaf bool) string {
   // If peer.service tag is present, use it directly as it is a reliable indicator of the service name.
   if peerService, found := getTagValue(span, "peer.service"); found {
      serviceName := peerService
      if isLeaf {
         serviceName += "-leaf"
      }
      return serviceName
   }
   ...
}
```